WindowsFormsSample/WindowsFormsSample/ChatForm.cs

```diff
@@ -0,0 +1,132 @@
+using System;
+using System.Drawing;
+using System.Windows.Forms;
+using Microsoft.AspNetCore.SignalR.Client;
+
+namespace WindowsFormsSample
+{
+    public partial class ChatForm : Form
+    {
+        private HubConnection _connection;
+
+        public ChatForm()
+        {
+            InitializeComponent();
+        }
+
+        private void ChatForm_Load(object sender, EventArgs e)
+        {
+            addressTextBox.Focus();
+        }
+
+        private void addressTextBox_Enter(object sender, EventArgs e)
+        {
+            AcceptButton = connectButton;
+        }
+
+        private async void connectButton_Click(object sender, EventArgs e)
+        {
+            UpdateState(connected: false);
+
+            _connection = new HubConnectionBuilder()
+                .WithUrl(addressTextBox.Text)
+                .Build();
+
+            _connection.On<string>("Send", OnSend);
+
+            Log(Color.Gray, "Starting connection...");
+            try
+            {
+                await _connection.StartAsync();
+            }
+            catch (Exception ex)
+            {
+                Log(Color.Red, ex.ToString());
+                return;
+            }
+
+            Log(Color.Gray, "Connection established.");
+
+            UpdateState(connected: true);
+
+            messageTextBox.Focus();
+        }
+
+        private async void disconnectButton_Click(object sender, EventArgs e)
+        {
+            Log(Color.Gray, "Stopping connection...");
+            try
+            {
+                await _connection.StopAsync();
+            }
+            catch (Exception ex)
+            {
+                Log(Color.Red, ex.ToString());
+            }
+
+            Log(Color.Gray, "Connection terminated.");
+
+            UpdateState(connected: false);
+        }
+
+        private void messageTextBox_Enter(object sender, EventArgs e)
+        {
+            AcceptButton = sendButton;
+        }
+
+        private async void sendButton_Click(object sender, EventArgs e)
+        {
+            try
+            {
+                await _connection.InvokeAsync("Send", messageTextBox.Text);
+            }
+            catch (Exception ex)
+            {
+                Log(Color.Red, ex.ToString());
+            }
+        }
+
+        private void UpdateState(bool connected)
+        {
+            disconnectButton.Enabled = connected;
+            connectButton.Enabled = !connected;
+            addressTextBox.Enabled = !connected;
+
+            messageTextBox.Enabled = connected;
+            sendButton.Enabled = connected;
+        }
+
+        private void OnSend(string message)
+        {
+            Log(Color.Black, message);
+        }
+
+        private void Log(Color color, string message)
+        {
+            messagesList.Items.Add(new LogMessage(color, message));
```

Comment:
Uhh @anurse how does this work when you're not on the sync context?

Reply:
That's a good point. OnSend should call Log in WinForms dispatcher

Reply:
I want to do more thinking about this. Here was the original discussion https://github.com/SignalR/SignalR/issues/972

Reply:
Well it works perfectly... I can update the the sample to put it on the dispatcher though.

Reply:
![https://media1.tenor.com/images/1359ccb442c9ed7aa8eef1df01935591/tenor.gif?itemid=7551542](https://media1.tenor.com/images/1359ccb442c9ed7aa8eef1df01935591/tenor.gif?itemid=7551542)

It doesn't throw and that was good enough for me at this point.